tools/src/wyvern/target/corewyvernIL/type/RefinementType.java

```diff
@@ -67,22 +72,52 @@ public RefinementType(List<ValueType> typeParams, ValueType base, HasLocation ha
                 // for debugging
                 base.checkWellFormed(ctx);
 
-                ToolError.reportError(ErrorMessage.CANNOT_APPLY_TYPE_PARAMETERS, this.getLocation(), base.toString());
+                ToolError.reportError(ErrorMessage.CANNOT_APPLY_GENERIC_ARGUMENTS, getLocation(), base.toString());
             }
-            int index = 0;
-            int declCount = st.getDeclTypes().size();
-            for (ValueType vt : typeParams) {
-                // advance the index to an AbstractTypeMember
-                while (index < declCount && !(st.getDeclTypes().get(index) instanceof AbstractTypeMember)) {
-                    index++;
+
+            final List<DeclType> stDeclTypes = st.getDeclTypes();
+            final Iterator<GenericArgument> genericArgumentIterator = genericArguments.iterator();
+
+            // Instantiate (make concrete) the abstract types with the arguments that are passed in
+            for (DeclType dt : stDeclTypes) {
+                // Generic arguments are used up
+                if (!genericArgumentIterator.hasNext()) {
```

Comment:
Can there be more generic arguments than `dt`s? If so, it needs to be handled; otherwise, please add a comment explaining why it can't be the case.

Reply:
If I'm understanding your concern correctly, I believe the next part of the code (starting with `// Too many generic arguments!`) handles this case.